From b70f51cdde14c774cd00cc5580d92f774c9d94bc Mon Sep 17 00:00:00 2001
From: "Chen, Yu" <yu.y.chen@intel.com>
Date: Tue, 1 Mar 2022 17:08:09 +0800
Subject: [PATCH] glsl/ast: Allow redeclaration of gl_Position

and gl_PointSize with different precision qualifier.

In shaders using version 300 and when extension EXT_separate_shader_objects is required,
redeclaration of gl_Position and gl_PointSize with different precision qualifier is allowed.
This fix can solve the black screen issue of game FIFA caused by shader compile failure.

Signed-off-by: Chen, Yu <yu.y.chen@intel.com>
Change-Id: Ied2e31245de414e224de4bf6dbaaa5906c242bad
---
 src/compiler/glsl/ast_to_hir.cpp | 23 ++++++++++++++++++++++-
 1 file changed, 22 insertions(+), 1 deletion(-)

diff --git a/src/compiler/glsl/ast_to_hir.cpp b/src/compiler/glsl/ast_to_hir.cpp
index 2a93a7fd401..57cc4923573 100644
--- a/src/compiler/glsl/ast_to_hir.cpp
+++ b/src/compiler/glsl/ast_to_hir.cpp
@@ -4480,7 +4480,28 @@ get_variable_being_redeclared(ir_variable **var_ptr, YYLTYPE loc,
       /* Allow verbatim redeclarations of built-in variables. Not explicitly
        * valid, but some applications do it.
        */
-   } else {
+   }
+   else if (state->is_version(140, 300) &&
+            state->has_separate_shader_objects() &&
+            (strcmp(var->name, "gl_Position") == 0 || strcmp(var->name, "gl_PointSize") == 0))
+   {
+      /* According to the EXT_separate_shader_objects spec:
+       *
+       *    "The following vertex shader outputs may be redeclared at global scope to
+       *    specify a built-in output interface, with or without special qualifiers:
+       *    gl_Position
+       *    gl_PointSize
+       *    When compiling shaders using either of the above variables, both such
+       *    variables must be redeclared prior to use.  ((Note:  This restriction
+       *    applies only to shaders using version 300 that enable the
+       *    EXT_separate_shader_objects extension; shaders not enabling the
+       *    extension do not have this requirement.))  A separable program object
+       *    will fail to link if any attached shader uses one of the above variables
+       *    without redeclaration."
+       */
+      earlier->data.precision = var->data.precision;
+   }
+   else {
       _mesa_glsl_error(&loc, state, "`%s' redeclared", var->name);
    }
 
-- 
2.33.1

